drivers/i2c/i2c_rcar.c

```diff
@@ -19,186 +19,207 @@ LOG_MODULE_REGISTER(i2c_rcar);
+		return i2c_rcar_wait_for_state(dev, RCAR_I2C_MAT | RCAR_I2C_MDE);
 	}
-	if (timeout == 100) {
-		return -ETIMEDOUT;
-	}
-
-	/* Check NAK */
-	if(i2c_rcar_read(config, RCAR_I2C_ICMSR) & RCAR_I2C_ICMSR_MNR)
-		return -EIO;
-
-	return 0;
 }
 
-static int i2c_rcar_transfer_msg(const struct device *dev, struct i2c_msg *msg){
+static int i2c_rcar_transfer_msg(const struct device *dev, struct i2c_msg *msg)
+{
 	const struct i2c_rcar_cfg *config = DEV_I2C_CFG(dev);
-	uint32_t icmcr = RCAR_I2C_ICMCR_MDBS | RCAR_I2C_ICMCR_MIE;
-	uint32_t i;
-	uint16_t timeout = 0;
+	uint32_t i, reg;
+	int ret;
+
+	if ((msg->flags & I2C_MSG_RW_MASK) == I2C_MSG_READ) {
+		/* Reading as master */
+		i2c_rcar_write(config, RCAR_I2C_ICMCR, RCAR_I2C_ICMCR_MASTER);
 
-	if ((msg->flags & I2C_MSG_RW_MASK) == I2C_MSG_READ){
 		for (i = 0; i < msg->len; i++) {
 			if (msg->len - 1 == i) {
-				icmcr |= RCAR_I2C_ICMCR_FSB;
+				i2c_rcar_write(config, RCAR_I2C_ICMCR, RCAR_I2C_ICMCR_MASTER | RCAR_I2C_ICMCR_FSB);
 			}
 
-			i2c_rcar_write(config, RCAR_I2C_ICMCR, icmcr);
-			i2c_rcar_write(config, RCAR_I2C_ICMSR, (uint32_t)~RCAR_I2C_ICMSR_MDR);
+			/* Start data reception */
+			reg = i2c_rcar_read(config, RCAR_I2C_ICMSR);
+			reg &= ~RCAR_I2C_MDR;
+			i2c_rcar_write(config, RCAR_I2C_ICMSR, reg);
 
-			/* Wait for a data to be received*/
-			while (!(i2c_rcar_read(config, RCAR_I2C_ICMSR) & RCAR_I2C_ICMSR_MDR) && (timeout<100)){
-				k_busy_wait(USEC_PER_MSEC);
-				timeout++;
-			}
-			if (timeout == 100) {
-				return -ETIMEDOUT;
+			/* Wait for data to be received */
+			ret = i2c_rcar_wait_for_state(dev, RCAR_I2C_MDR);
+			if (ret != 0) {
+				return ret;
 			}
 
 			msg->buf[i] = i2c_rcar_read(config, RCAR_I2C_ICRXD_ICTXD) & 0xff;
 		}
-
-		i2c_rcar_write(config, RCAR_I2C_ICMSR, (uint32_t)~RCAR_I2C_ICMSR_MDR);
-	}
-	else {
+	} else {
+		/* Writting as master */
 		for (i = 0; i < msg->len; i++) {
 			i2c_rcar_write(config, RCAR_I2C_ICRXD_ICTXD, msg->buf[i]);
-			i2c_rcar_write(config, RCAR_I2C_ICMCR, icmcr);
-			i2c_rcar_write(config, RCAR_I2C_ICMSR, (uint32_t)~RCAR_I2C_ICMSR_MDE);
 
-			while (!(i2c_rcar_read(config, RCAR_I2C_ICMSR) & RCAR_I2C_ICMSR_MDE) && (timeout<100)){
-				k_busy_wait(USEC_PER_MSEC);
-				timeout++;
-			}
-			if (timeout == 100) {
-				return -ETIMEDOUT;
+			i2c_rcar_write(config, RCAR_I2C_ICMCR, RCAR_I2C_ICMCR_MASTER);
+
+			/* Start data transmission */
+			reg = i2c_rcar_read(config, RCAR_I2C_ICMSR);
+			reg &= ~RCAR_I2C_MDE;
+			i2c_rcar_write(config, RCAR_I2C_ICMSR, reg);
+
+			/* Wait for all data to be transmitted */
+			ret = i2c_rcar_wait_for_state(dev, RCAR_I2C_MDE);
+			if (ret != 0) {
+				return ret;
 			}
 		}
-
-		i2c_rcar_write(config, RCAR_I2C_ICMSR, (uint32_t)~RCAR_I2C_ICMSR_MDE);
-		icmcr |= RCAR_I2C_ICMCR_FSB;
-		i2c_rcar_write(config, RCAR_I2C_ICMCR, icmcr);
 	}
 
-	return i2c_rcar_finish(dev);
+	/* Finish the transfer */
+	if ((msg->flags & I2C_MSG_STOP) == I2C_MSG_STOP) {
+		return i2c_rcar_finish(dev);
+	} else {
+		return 0;
+	}
```

Comment:
```suggestion
	if ((msg->flags & I2C_MSG_STOP) == I2C_MSG_STOP) {
		ret = i2c_rcar_finish(dev);
	} 
	return ret;
```

Reply:
Are we so assuming that we are returning last call for "wait for state" return if the message is not a stopping one ?